### PR TITLE
Add representation-specific entry map in Production and Consumption.

### PR DESCRIPTION
**Pull request by msporny**
This PR attempts to address @peacekeeper's long standing concern wrt. production and consumption and representation-specific entries by using the proposal for Option 3.3 in issue https://github.com/w3c/did-core/issues/664#issuecomment-780647353. This is also an attempt to get PRs #596 and #597 unstuck. 

Specifically, it defines the input and return parameters for conforming producers and consumers and splits the representation-specific entries into their own map that can be serialized and returned along with the DID Document data model. It adds a requirement that all representations detect known representation-specific entries, which is required to cleanly create the two classes of information. The change is substantive.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/pull/679.html" title="Last updated on Feb 22, 2021, 2:47 PM UTC (bd42543)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/679/4fa8635...bd42543.html" title="Last updated on Feb 22, 2021, 2:47 PM UTC (bd42543)">Diff</a>

### CODE REVIEW

File: index.html
Comment (msporny) [inline index.html:2482]: I had played around with alternative language here that effectively said: "any representation specific entries that could appear in the representation" -- which basically boils down to just `@context`... `%YAML` and `xmlns` are highly unlikely to appear in JSON and CBOR.

File: index.html
Comment (msporny) [inline index.html:2827]:
```suggestion
representation-specific entries <a data-cite="INFRA#maps">map</a> MUST be
```

**Review by OR13 — CHANGES_REQUESTED** [review]
I would prefer to see abstract function signatures for consume and produce.

something like:

``` 
produce = (didDocument, representation-entries, representation-options) => didDocumentByteStream
consume = (didDocumentByteStream) => didDocument, representation-entries
```

I don't understand why we define abstract function for resolution, and then not for production and consumption.

`representation-entries` and `representation-options` need to be defined next to `didDocumnt` and `didDocumentByteStream`

If an issue is added pointing to this comment, I am happy to approve and do the PR myself.

File: index.html
Comment (TallTed) [inline index.html:2454]: After reading this over several times, I *think* I've figured out a small
set of changes that make it clearer after fewer readings, hopefully down
to one read for sufficient clarity to act on.
```suggestion
A <a>conforming producer</a> MUST serialize all entries in the <a>DID
document</a> <a href="#data-model">data model</a>, and the relevant
entries in the representation-specific entries <a
data-cite="INFRA#maps">map</a>, that do not have explicit processing
rules for the <a>representation</a> being produced, using only the
<a>representation</a>'s data type processing rules; and return the
serialization after the <a>production</a> process completes.
```

File: index.html
Comment (TallTed) [inline index.html:2457]:
```suggestion
A <a>conforming producer</a> MUST return the Media Type <a
```
_MIME Types_ became _Media Types_ with [RFC 1590](https://tools.ietf.org/html/rfc1590) in 1994.  (There are several, if not many, places this change is needed.)

File: index.html
Comment (TallTed) [inline index.html:2470]:
```suggestion
Media Type <a data-cite="INFRA#string">string</a> as input into
```
or "media type" ... Capitalization may be important for clarity.

File: index.html
Comment (TallTed) [inline index.html:2828]:
```suggestion
representation-specific entries <a data-cite="INFRA#maps">map</a> MUST be
serialized to the JSON-LD <a>representation</a> according to the JSON
```

File: index.html
Comment (TallTed) [reply index.html:2827]: `MUST be MUST be` becomes `MUST be be` ... so take my suggestion below instead

**Comment by msporny** [comment]
> I don't understand why we define abstract function for resolution, and then not for production and consumption.

With my Editor's hat firmly on: We never should have defined abstract functions for resolution using our own bespoke abstract function language... I cannot underscore how terrible the half-measure that is the Resolution section is -- we're bending over backwards and doing really strange stuff there to half-define something. We should have either just kept all of it to a testable algorithm in prose, like https://www.w3.org/TR/json-ld11-api/#expansion-algorithm does... or actually defined a concrete set of functions using Javascript/Typescript/WebIDL. What we have is a weird half-measure and I'm reluctant to double-down on that mistake.

I did start to write that up initially, and then decided against it because...

> representation-options

We have no way to realize this parameter in the way the Resolution section is written (which again, is strange and bespoke and isn't clearly specified for anything that is not INFRA->JSON... like, you know, CBOR). None of the data types in the data model or INFRA support it. Consider the use case of injecting encoded CBOR values into the root CBOR map... integers as keys -- can't do it... no concrete representation in INFRA for CBOR. So the best I could do is use prose to say that you can take other options and those options can be used to affect production. We're not testing MAY statements... that's the only way I could get to making representation-options a possibility.

It's not that it's impossible to write the spec up with bespoke abstract function language... but it makes it worse, IMHO, not better. All that said, I'll draft a PR that does the abstract function thing (even though I'm totally against the concept... what we have now is testable and flexible enough to support representation-options). If I write it up like we did for Resolution, representation-options are still not going to be a thing because we have no way of expressing them.

File: index.html
Comment (msporny) [reply index.html:2457]: Yes, I'm going to do this in a big spec-wide PR in one go. Rejecting your change suggestion, even though it's correct, want to bundle all those MIME type -> Media Type changes together /after/ the spec has settled.

File: index.html
Comment (msporny) [reply index.html:2470]: Rejecting per https://github.com/w3c/did-core/pull/679/files#r579329532

File: index.html
Comment (msporny) [reply index.html]: Done.

File: index.html
Comment (TallTed) [reply index.html:2458]: Fair ... though I don't know what would be lost by doing some of these changes sooner, and think that having a mix of MIME & Media would be more likely to flag the need to replace the former than having all MIME would.  I'll count on you having a list of such "will do these later in batches" changes.

File: index.html
Comment (msporny) [reply index.html:2470]: Fixed in d915bea303b36c00ec59b27b8ac10845da828fea

File: index.html
Comment (msporny) [reply index.html:2457]: Fixed in d915bea303b36c00ec59b27b8ac10845da828fea.

File: index.html
Comment (TallTed) [inline index.html:1140]:
```suggestion
high-level description of the data model, descriptions of the ways different
types of properties are expressed in the data model, and instructions for
extending the data model.
```

File: index.html
Comment (brentzundel) [inline index.html:2827]:
```suggestion
representation-specific entries <a data-cite="INFRA#maps">map</a> MUST
```

**Review by brentzundel — APPROVED** [review]
I believe this PR includes changes from PR #678. Not sure if that was deliberate.

**Comment by OR13** [comment]
> We should have either just kept all of it to a testable algorithm in prose, like https://www.w3.org/TR/json-ld11-api/#expansion-algorithm does... or actually defined a concrete set of functions using Javascript/Typescript/WebIDL.

I agree... it's not too late to fix this ambiguity. I personally find function definitions more useful than english sentences....

```
didDocument =  « 
didDocumentEntries: INFRA<map>, 
didDocumentRepresentationEntries: INFRA<map> 
»

resolve(
did: INFRA<string>, 
resolutionOptions: INFRA<map>
) →
   « 
   didResolutionMetadata: INFRA<map>, 
   didDocument: INFRA<map>, 
   didDocumentMetadata: INFRA<map> 
   »
   
resolveRepresentation(
did: INFRA<string>, 
resolutionOptions: INFRA<map>
) →
   « 
   didResolutionMetadata: INFRA<map>, 
   didDocumentStream: INFRA<ByteStream>, 
   didDocumentMetadata: INFRA<map> 
   »
      
consume(
contentType: INFRA<string>, 
didDocumentStream: INFRA<ByteStream>
) →
  « 
  didDocumentEntries: INFRA<map>, 
  representationEntries: INFRA<map> 
  »
  
produce(
contentType: INFRA<string>, 
didDocumentEntries: INFRA<map>, 
didDocumentRepresentationEntries: INFRA<map>
) →
  « 
  didDocumentStream: INFRA<ByteStream> 
  »
```

**Comment by msporny** [comment]
@OR13 wrote:
> I personally find function definitions far more useful than english sentences....

Sure, I do as well, and W3C has a formal language in which to do that... it's called WebIDL... which people objected to using.

Instead, we've got a interface definition language that we invented on top of INFRA -- totally bespoke, no one understands it except for us... and we're using that to define the interfaces in the specification, which BTW, we're totally not chartered to do. The WG decided to look the other way wrt. the Resolution section on that particular point.

**Comment by msporny** [comment]
This PR is blocked waiting on a review from: @peacekeeper and @jricher

File: index.html
Comment (peacekeeper) [inline index.html:2481]: Perhaps we want to say "MUST detect any representation-specific entry known to the representation being consumed", and "SHOULD/MAY detect representation-specific entries across all other known representations"? Anyway I'm also fine with this as-is.

**Review by peacekeeper — APPROVED** [review]
> bespoke abstract function language

I don't have the same spec writing experience you do, so I'm not sure what other specs do when you want to define an interface (with inputs and outputs) for a process that can have many concrete realizations.

We can still add examples for WebIDL (https://github.com/w3c/did-core/pull/587) and TypeScript (https://github.com/w3c/did-core/pull/591), I think that would be useful as long as it's clear that they are just examples, not the authoritative definitions.

I also like @OR13 's proposal in https://github.com/w3c/did-core/pull/679#issuecomment-782481740 to add function definitions for `produce()` and `consume()`. We can again define them in an abstract way with concrete examples in WebIDL and TypeScript.

> close to going outside of our charter

My preference at the Amsterdam F2F was also to not include DID Resolution in DID Core, but since then we've operated under the assumption that we want to define the "contract" (without its implementation) of DID Resolution (and DID URL Dereferencing).

**Comment by msporny** [comment]
@peacekeeper wrote:
> My preference at the Amsterdam F2F was also to not include DID Resolution in DID Core, but since then we've operated under the assumption that we want to define the "contract" (without its implementation) of DID Resolution (and DID URL Dereferencing).

Then I think it's time to revisit that decision given new information, as that would be my preference as well. Perhaps we do this if it becomes clear that volunteers for writing the tests and two independent implementations are not forthcoming by the end of the first Candidate Recommendation phase.

File: index.html
Comment (msporny) [reply index.html:2481]: Yes, I had considered that and decided not to go that route because it makes detection of other representation-specific entries optional, which would ultimately lead to interoperability issues on a critical feature needed to solve the problem we're trying to solve. Making this behavior mandatory removes the possibility of non-interoperability at the representation-specific entries data model.

**Comment by msporny** [comment]
Substantive, multiple reviews, changes requested and made, no blocking objections (but a standing request from @OR13 exists to possibly define abstract functions), merging.

**Pull request merged**
